File: app/src/main/java/com/niiadotei/storehouse/ui/customers/CustomersViewModel.java
package com.niiadotei.storehouse.ui.customers;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.app.Dialog;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.text.InputType;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;

import com.niiadotei.storehouse.R;
import com.niiadotei.storehouse.data.DatabaseHelper;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class CustomersViewModel extends RecyclerView.Adapter<CustomersViewModel.ViewHolder> implements Filterable {
    Fragment fragment;
    JSONArray jsonArray;

    DatabaseHelper databaseHelper;

    JSONArray filteredJsonArray;

    public FilteredViewHolder filteredViewHolder;

    public CustomersViewModel(Fragment fragment, JSONArray jsonArray) {
        this.fragment = fragment;
        this.jsonArray = jsonArray;

        filteredJsonArray = new JSONArray();
    }

    public void updateArray(JSONArray jsonArray) {
        this.jsonArray = jsonArray;
    }

    @NonNull
    @Override
    public CustomersViewModel.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_customer, parent, false);

        databaseHelper = new DatabaseHelper(view.getContext());

        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull CustomersViewModel.ViewHolder holder, int position) {
        try {
            JSONObject jsonObject = jsonArray.getJSONObject(position);

            holder.customerNameTextView.setText(jsonObject.getString("name"));

            String contact = "Contact: " + jsonObject.getString("phone");
            holder.phoneNumberTextView.setText(contact);

            String location = "Location: " + jsonObject.getString("location");
            holder.locationTextView.setText(location);

            String date = "Date added: " + jsonObject.getString("date");
            holder.dateTextView.setText(date);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        holder.itemView.setOnClickListener(view -> {
            try {
                JSONObject jsonObject = jsonArray.getJSONObject(holder.getAdapterPosition());

                String idFromArray = jsonObject.getString("id");
                String nameFromArray = jsonObject.getString("name");
                String phoneNumberFromArray = jsonObject.getString("phone");
                String locationFromArray = jsonObject.getString("location");

                Dialog dialog = new Dialog(fragment.getActivity());
                dialog.getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
                dialog.setContentView(R.layout.dialog_customer);
                dialog.show();

                EditText nameEditText = dialog.findViewById(R.id.customerNameEditText);
                EditText phoneEditText = dialog.findViewById(R.id.phoneNumberEditText);
                EditText locationEditText = dialog.findViewById(R.id.locationEditText);
                Button updateCustomerButton = dialog.findViewById(R.id.addCustomerButton);

                nameEditText.setText(nameFromArray);
                nameEditText.setInputType(InputType.TYPE_NULL);
                nameEditText.setTextIsSelectable(false);

                phoneEditText.setText(phoneNumberFromArray);
                locationEditText.setText(locationFromArray);
                updateCustomerButton.setText(R.string.update_button);

                updateCustomerButton.setOnClickListener(view1 -> {
                    String updatedPhoneNumber = phoneEditText.getText().toString().trim();
                    String updatedLocation = locationEditText.getText().toString().trim();

                    databaseHelper.updateCustomer(idFromArray, updatedPhoneNumber, updatedLocation);
                    updateArray(databaseHelper.getCustomerArray());
                    notifyItemChanged(holder.getAdapterPosition());
                    dialog.dismiss();
                });

            } catch (JSONException e) {
                e.printStackTrace();
            }
        });

        holder.itemView.setOnLongClickListener(view -> {
            int getPosition = holder.getAdapterPosition();

            try {
                JSONObject jsonObject = jsonArray.getJSONObject(getPosition);

                String stringID = jsonObject.getString("id");

                AlertDialog.Builder builder = new AlertDialog.Builder(fragment.getActivity());
                builder.setTitle("Confirm");
                builder.setMessage("Are you sure you want to delete this customer?");

                builder.setPositiveButton("Yes", (dialogInterface, i) -> {
                    databaseHelper.deleteCustomer(stringID);
                    jsonArray.remove(getPosition);
                    notifyItemRemoved(getPosition);
                    notifyItemRangeChanged(getPosition, jsonArray.length());
                });

                builder.setNegativeButton("No", ((dialogInterface, i) -> dialogInterface.dismiss()));

                builder.show();
            } catch(JSONException e) {
                e.printStackTrace();
            }
            return true;
        });
    }

    @Override
    public int getItemCount() {
        return jsonArray.length();
    }

    @Override
    public Filter getFilter() {
        if (filteredViewHolder == null) {
            filteredJsonArray = jsonArray;

            filteredViewHolder = new CustomersViewModel.FilteredViewHolder(this, filteredJsonArray);
        }

        return filteredViewHolder;
    }

    public JSONArray getFilteredJsonArray(String name) {
        return databaseHelper.getFilteredCustomerArray(name);
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        TextView customerNameTextView;
        TextView phoneNumberTextView;
        TextView locationTextView;
        TextView dateTextView;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);

            customerNameTextView = itemView.findViewById(R.id.customerNameTextView);
            phoneNumberTextView = itemView.findViewById(R.id.customerPhoneNumberTextView);
            locationTextView = itemView.findViewById(R.id.locationTextView);
            dateTextView = itemView.findViewById(R.id.customerDateAddedTextView);
        }
    }

    private static class FilteredViewHolder extends Filter {
        private final CustomersViewModel customersViewModel;
        private final JSONArray jsonArray;
        private JSONArray filteredJsonArray;

        private FilteredViewHolder(CustomersViewModel customersViewModel, JSONArray jsonArray) {
            this.customersViewModel = customersViewModel;
            this.jsonArray = jsonArray;
            this.filteredJsonArray = new JSONArray();

        }

        @Override
        protected FilterResults performFiltering(CharSequence charSequence) {
            final FilterResults results = new FilterResults();

            if (charSequence.length() == 0) {
                filteredJsonArray = jsonArray;
            } else {
                final String filterPattern = charSequence.toString().trim();

                filteredJsonArray = customersViewModel.getFilteredJsonArray(filterPattern);
            }

            results.values = filteredJsonArray;
            return results;
        }

        @SuppressLint("NotifyDataSetChanged")
        @Override
        protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
            customersViewModel.updateArray((JSONArray) filterResults.values);
            customersViewModel.notifyDataSetChanged();
        }
    }
}

File: app/src/main/java/com/niiadotei/storehouse/ui/customers/CustomersViewModel.kt
package com.niiadotei.storehouse.ui.customers

import android.annotation.SuppressLint

import android.app.AlertDialog
import android.app.Dialog

import android.content.DialogInterface
import android.content.Intent

import android.graphics.Color
import android.graphics.drawable.ColorDrawable

import android.text.InputType

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import android.widget.*

import androidx.fragment.app.Fragment

import androidx.recyclerview.widget.RecyclerView

import com.niiadotei.storehouse.R
import com.niiadotei.storehouse.data.DatabaseHelper
import com.niiadotei.storehouse.ui.customers.purchases.CustomerPurchasesActivity

import org.json.JSONArray
import org.json.JSONException

class CustomersViewModel(var fragment: Fragment, var jsonArray: JSONArray) : RecyclerView.Adapter<CustomersViewModel.ViewHolder>(), Filterable {
    private lateinit var databaseHelper: DatabaseHelper

    private var filteredJsonArray: JSONArray

    init {
        filteredJsonArray = JSONArray()
    }

    fun updateArray(jsonArray: JSONArray) {
        this.jsonArray = jsonArray
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_customer, parent, false)
        databaseHelper = DatabaseHelper(view.context)

        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        try {
            val jsonObject = jsonArray.getJSONObject(position)

            holder.customerNameTextView.text = jsonObject.getString("name")

            val contact = "Contact: " + jsonObject.getString("phone")
            holder.phoneNumberTextView.text = contact

            val location = "Location: " + jsonObject.getString("location")
            holder.locationTextView.text = location

            val date = "Date added: " + jsonObject.getString("date")
            holder.dateTextView.text = date
        } catch (e: JSONException) {
            e.printStackTrace()
        }

        holder.itemView.setOnClickListener { view: View ->
            try {
                val jsonObject = jsonArray.getJSONObject(position)

                val customer = jsonObject.getString("name") + ", " + jsonObject.getString("phone")

                val intent = Intent(view.context, CustomerPurchasesActivity::class.java)
                intent.putExtra("customer", customer)

                view.context.startActivity(intent)
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }

        holder.itemView.setOnLongClickListener {
            val getPosition = holder.adapterPosition

            try {
                val jsonObject = jsonArray.getJSONObject(getPosition)

                val stringID = jsonObject.getString("id")

                val builder = AlertDialog.Builder(fragment.activity)
                builder.setTitle(R.string.confirm_title)
                builder.setMessage(R.string.delete_customer_message)
                builder.setPositiveButton(R.string.positive_button) { _: DialogInterface?, _: Int ->
                    databaseHelper.deleteCustomer(stringID)
                    jsonArray.remove(getPosition)
                    notifyItemRemoved(getPosition)
                    notifyItemRangeChanged(getPosition, jsonArray.length())
                }
                builder.setNegativeButton(R.string.negative_button) { _: DialogInterface?, _: Int ->
                    try {
                        val idFromArray = jsonObject.getString("id")
                        val nameFromArray = jsonObject.getString("name")
                        val phoneNumberFromArray = jsonObject.getString("phone")
                        val locationFromArray = jsonObject.getString("location")

                        val dialog = Dialog(fragment.requireActivity())
                        dialog.window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
                        dialog.setContentView(R.layout.dialog_customer)
                        dialog.show()

                        val nameEditText = dialog.findViewById<EditText>(R.id.customerNameEditText)
                        val phoneEditText = dialog.findViewById<EditText>(R.id.phoneNumberEditText)
                        val locationEditText = dialog.findViewById<EditText>(R.id.locationEditText)
                        val updateCustomerButton = dialog.findViewById<Button>(R.id.addCustomerButton)

                        nameEditText.setText(nameFromArray)
                        nameEditText.inputType = InputType.TYPE_NULL
                        nameEditText.setTextIsSelectable(false)

                        phoneEditText.setText(phoneNumberFromArray)

                        locationEditText.setText(locationFromArray)

                        updateCustomerButton.setText(R.string.update_button)
                        updateCustomerButton.setOnClickListener {
                            val updatedPhoneNumber = phoneEditText.text.toString().trim { it <= ' ' }
                            val updatedLocation = locationEditText.text.toString().trim { it <= ' ' }

                            databaseHelper.updateCustomer(idFromArray, updatedPhoneNumber, updatedLocation)

                            updateArray(databaseHelper.customerArray)
                            notifyItemChanged(holder.adapterPosition)
                            dialog.dismiss()
                        }
                    } catch (e: JSONException) {
                        e.printStackTrace()
                    }
                }
                builder.show()
            } catch (e: JSONException) {
                e.printStackTrace()
            }

            true
        }
    }

    override fun getItemCount(): Int {
        return jsonArray.length()
    }

    override fun getFilter(): Filter {
        return object: Filter() {
            override fun performFiltering(charSequence: CharSequence): FilterResults {
                val results = FilterResults()

                filteredJsonArray = if (charSequence.isEmpty()) {
                    jsonArray
                } else {
                    val filterPattern = charSequence.toString().trim { it <= ' ' }
                    getFilteredJsonArray(filterPattern)
                }

                results.values = filteredJsonArray
                return results
            }

            @SuppressLint("NotifyDataSetChanged")
            override fun publishResults(charSequence: CharSequence, filterResults: FilterResults) {

                if (charSequence.isEmpty()) {
                    updateArray(jsonArray)
                } else {
                    updateArray(filterResults.values as JSONArray)
                }

                notifyDataSetChanged()
            }
        }
    }

    fun getFilteredJsonArray(name: String?): JSONArray {
        return databaseHelper.getFilteredCustomerArray(name!!)
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var customerNameTextView: TextView
        var phoneNumberTextView: TextView
        var locationTextView: TextView
        var dateTextView: TextView

        init {
            customerNameTextView = itemView.findViewById(R.id.customerNameTextView)
            phoneNumberTextView = itemView.findViewById(R.id.customerPhoneNumberTextView)
            locationTextView = itemView.findViewById(R.id.locationTextView)
            dateTextView = itemView.findViewById(R.id.customerDateAddedTextView)
        }
    }

}